### ISSUE TITLE
WCAG support

### ISSUE THREAD

**Issue by ellinge** [dowjones/react-dropdown-tree-select#216]
There's currently no support for keyboard navigation and no ARIA-attributes present.
Here are some guidelines one should consider complying with WCAG:
https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html

Here are some of the highlights on attributes one should consider implementing:
- role
- aria-haspopup
- aria-expanded
- aria-selected

**Comment by mrchief** [comment]
Thanks for opening this. Having WCAG support would be great. I believe there are no clear guidelines for a mixed or combo control like this.

This component essentially is a mix of:

 - an input
 - a dropdown 
 - a TreeView
 - label + checkboxes (each tree node)
 - tags (selected tree nodes that have text plus a "x" button)
 - node action item(s) - these are optional but essentially they are just buttons

To fully implement WCAG in a meaningful way, we'd need the following:

#### To add ARIA attributes
 - [ ] Determine proper attributes to apply
 - [ ] PR for adding the attributes
 - [ ] Automated tests to ensure compliance (preferably integrated with CI)

#### To add Keyboard navigation
 - [ ] Figure out what Keyboard navigation should look like with all possible variations (e.g., with or without checkboxes, with or without action items, scrolling, adding/removing/navigating tags, etc.)
 - [ ] Handle scroll virtualization 
 - [ ] PR for Adding keyboard navigation capabilities
 - [ ] Automated tests to ensure compliance (preferably integrated with CI)

As you can see, the scope can be quite broad for getting it right a 100%, but I think we can probably add stuff incrementally in multiple PRs. 

I'll start looking into it, and I'm open to ideas, suggestions, and PRs for this.

**Comment by ellinge** [comment]
Did some first tries to include a11y-testing through axe:
https://github.com/ellinge/react-dropdown-tree-select/tree/feat/%23216

It only detects missing labeling for search input though. Not sure it will detect our scenario/needs. Since some dom-nodes are not rendered by React, it propably cannot detect that we expand and collapse child nodes. Not sure of it's capabilties though. But it probably will be able to validate the attributes we apply I guess.

They also state
> There is limited support for JSDOM. We will attempt to make all rules compatible with JSDOM but where this is not possible, we recommend turning those rules off. Currently the color-contrast rule is known not to work with JSDOM.

Here is the rule list:
https://github.com/dequelabs/axe-core/blob/develop/doc/rule-descriptions.md

**Comment by ellinge** [comment]
There’s automated tests in place in #225 
Scroll virtulization? Do you mean replacing infinitescroll with a virtualized list/window instead?

**Comment by mrchief** [comment]
> Scroll virtualization

I meant scroll should work with arrow/pg up/dn keys...

**Comment by ellinge** [comment]
> 
> 
> > Scroll virtualization
> 
> I meant scroll should work with arrow/pg up/dn keys...

Alright, then that's in place of #225 also. A virtual window would be nice also and solve another open issue I think. Lots of potential nodes to render on page down 🐢

**Comment by mrchief** [comment]
>  A virtual window would be nice though

Do you mean something like react-window or similar? I've considered many in the past and most of them require a fixed height which is kind of a downer for me. It's an additional burden for the caller which I would like to avoid. The current solution seems like a good one though, is there something bothering you?

> Lots of potential nodes to render on page down

Page down should render the next page, similar to what happens when you reach the last item via mouse scroll (it renders the next set of 100 items, which is our page size).

> The audit for Chrome/axe doesn't solve more than validation of added aria-attributes for us, custom controls are open to interpretation. But Code climate seems to be aware of them already so perhaps they are redundant and should be removed.

I was wondering the same. Between code climate and our RFC assertions, we should be ok so yeah, remove axe if you prefer so.

**Comment by ellinge** [comment]
Realised since we set them dynamically, code climate won't detect them.
So I'll leave axe since that's at least "rendered", even if only virtually.

**Comment by ellinge** [comment]
Realized since we set them dynamically, code climate won't detect them.
So I'll leave axe since that's at least "rendered", even if only virtually.
